s3torchconnector/src/s3torchconnector/s3reader/dcp_optimized.py

```diff
@@ -0,0 +1,588 @@
+
+    def __init__(self) -> None:
+        self._segments: List[memoryview] = []  # memoryview segments
+        self._offsets: List[int] = []  # start offset (within the item) of each segment
+        self._lengths: List[int] = []  # length of each segment
+        self._size: int = 0  # total item length (sum of _lengths)
+        self._pos: int = 0  # current read position within the item
+        self._closed: bool = False
+
+    def append_view(self, view: memoryview) -> None:
+        """Append a memoryview segment (ignored if empty)."""
+        assert not self._closed, "Buffer is closed"
+
+        seg_len = len(view)
+        if seg_len == 0:
+            return
+        self._segments.append(view)
+        self._offsets.append(self._size)
+        self._lengths.append(seg_len)
+        self._size += seg_len
+
+    def close(self) -> None:
+        if not self._closed:
+            self._closed = True
+            self._segments.clear()
+
+    def seek(self, offset: int, whence: int = SEEK_SET, /) -> int:
+        assert isinstance(offset, int), f"integer expected, got {type(offset)!r}"
+
+        if whence == SEEK_SET:
+            new_pos = offset
+        elif whence == SEEK_CUR:
+            new_pos = self._pos + offset
+        elif whence == SEEK_END:
+            new_pos = self._size + offset
+        else:
+            raise ValueError(
+                "Seek must be passed io SEEK_CUR, SEEK_SET, or SEEK_END integers"
+            )
+
+        assert new_pos >= 0, f"negative seek value {new_pos}"
+
+        # Seeking past EOF is allowed.
+        self._pos = new_pos
+        return self._pos
+
+    def tell(self) -> int:
+        """Return the current pos position (like BytesIO.tell)."""
+        return self._pos
+
+    def read(self, size: Optional[int] = None) -> bytes:
+        assert size is not None, "Size cannot be None; full read is not supported"
+        assert size >= 0, "Size cannot be negative; full read is not supported"
+
+        # Fast path for read(4) at pos=0 (Optimizes pytorch/torch/serialization.py _is_zipfile())
+        if size == 4 and self._pos == 0 and self._lengths and self._lengths[0] >= 4:
+            self._pos = 4
+            # TODO: eliminating bytes() conversion can save ~3% time? Requires interface changes.
+            return bytes(self._segments[0][:4])
+
+        if size == 0:
+            return b""
+
+        # Pass implementation to readinto()
+        out = bytearray(size)
+        n = self.readinto(out)
+        return bytes(out) if n == size else memoryview(out)[:n].tobytes()
+
+    def readinto(self, buf) -> int:
+        dest = buf if isinstance(buf, memoryview) else memoryview(buf)
+        assert not dest.readonly, "writable buffer required"
+
+        dest_len = len(dest)
+        size = self._size
+        pos = self._pos
+
+        if dest_len == 0 or pos >= size:
+            return 0
+
+        # Cache to avoid repeated attribute calls
```

Comment:
This is very low level optimisation. Given you're using `__slots__`, doesn't that typically remove the cost of `self.` getattr calls?